### PR TITLE
C#: Improve performance of CFG split set computation

### PR DESCRIPTION
**Pull request by hvitved**
Rewrite the predicate `succSplits()` and the construction of the IPA type `TSplits`. The two are now mutually dependent, see more in the comment for the module `SuccSplits`.

I have tested manually on a few snapshots, including https://lgtm.com/projects/g/pickypg/runuo/ and https://github.com/dotnet/winforms, where the old implementation would timeout because of excessive CFG splitting. I will conduct a more thorough performance test, but I am quite confident that this should be a win on most snapshots.

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 1]
```diff
@@ -3814,150 +3730,212 @@ module ControlFlow {
...
-          split = succSplits.getASplit()
-        )
-        or
-        exists(SplitInternal predSplit |
-          succSplits3(pred, predSplits, TSplitsCons(predSplit, predSplitsRemaining), succ, succSplits, c) |
-          // Each predecessor split must be either passed over as a successor split,
-          // or must be left (possibly entering a new split)
-          predSplit.hasSuccessor(pred, succ, c) and
-          predSplit = succSplits.getASplit()
+        private predicate succSplits1b0(ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Completion c) {
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: A few comments in this predicate wouldn't go amiss.

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 2]
```diff
@@ -3814,150 +3730,212 @@ module ControlFlow {
...
+          exists(Reachability::SameSplitsBlock b |
+            succReachable(b, pred, predSplits, succ, c) |
+            (succ = b.getAnElement() implies succ = b) and
+            not exists(SplitInternal split | split.hasEntry(pred, succ, c))
+          )
+        }
+
+        // A `forall` over the splits in `predSplits`, but written using explicit recursion
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Would be good to document the `forall` invariant - which I presume is that all the splits are the same?

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 3]
```diff
@@ -3814,150 +3730,212 @@ module ControlFlow {
...
-      predicate succExitSplits(ControlFlowElement pred, Splits predSplits, Callable succ, SuccessorType t) {
-        exists(Reachability::SameSplitsBlock b, Completion c |
-          pred = b.getAnElement() |
-          b.isReachable(predSplits) and
-          t.matchesCompletion(c) and
-          succ = succExit(pred, c) and
-          forall(SplitInternal predSplit |
-            predSplit = predSplits.getASplit() |
-            predSplit.hasExit(pred, c)
+        private predicate succSplits2aux(ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Completion c) {
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Documentation for this predicate would be appreciated!

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 4]
```diff
@@ -3814,150 +3730,212 @@ module ControlFlow {
...
+          sk.appliesTo(succ) and
+          remaining = predSplits
+          or
+          exists(Splits mid, SplitInternal split |
+            succSplits2aNoneOfKindForall(pred, predSplits, succ, c, sk, mid) and
+            mid = TSplitsCons(split, remaining)
+          |
+            split.getKind() = any(SplitKind sk0 |
+              sk0 != sk and
+              (sk0.appliesTo(succ) or split.hasExit(pred, succ, c))
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: I think the condition `split.hasExit(pred, succ, c)` can be removed as it's already on line 3844.

Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Yes, I have removed it.

**Pull request merged**